history.go

```diff
@@ -142,6 +142,211 @@ type ReqHistoryDeleteProcessInstance struct {
+	// lteq - lower than or equal to;
+	// like.
+	// key and value may not contain underscore or comma characters.
+	TaskVariables []VariableFilterExpression `json:"taskVariables,omitempty"`
+	// Only include tasks that belong to process instances that have variables with certain values.Variable filtering expressions are comma-separated and are structured as follows:
+	// A valid parameter value has the form key_operator_value.key is the variable name, operator is the comparison operator to be used and value the variable value.
+	// Note: Values are always treated as String objects on server side.
+	//
+	// Valid operator values are: eq - equal to;
+	// neq - not equal to;
+	// gt - greater than;
+	// gteq - greater than or equal to;
+	// lt - lower than;
+	// lteq - lower than or equal to;
+	// like.
+	// key and value may not contain underscore or comma characters.
+	ProcessVariables []VariableFilterExpression `json:"processVariables,omitempty"`
+	// Match the variable name provided in taskVariables and processVariables case-insensitively. If set to true variableName and variablename are treated as equal.
+	VariableNamesIgnoreCase bool `json:"variableNamesIgnoreCase,omitempty"`
+	//Match the variable value provided in taskVariables and processVariables case-insensitively. If set to true variableValue and variablevalue are treated as equal.
+	VariableValuesIgnoreCase bool `json:"variableValuesIgnoreCase,omitempty"`
+	// Restrict to tasks with a historic identity link to the given user.
+	TaskInvolvedUser *string `json:"taskInvolvedUser,omitempty"`
+	// Restrict to tasks with a historic identity link to the given group.
+	TaskInvolvedGroup *string `json:"askInvolvedGroup,omitempty"`
+	// Restrict to tasks with a historic identity link to the given candidate user.
+	TaskHadCandidateUser *string `json:"taskHadCandidateUser,omitempty"`
+	// Restrict to tasks with a historic identity link to the given candidate group.
+	TaskHadCandidateGroup *string `json:"taskHadCandidateGroup,omitempty"`
+	// Only include tasks which have a candidate group. Value may only be true, as false is the default behavior.
+	WithCandidateGroups bool `json:"withCandidateGroups,omitempty"`
+	// Only include tasks which have no candidate group. Value may only be true, as false is the default behavior.
+	WithoutCandidateGroups bool `json:"withoutCandidateGroups,omitempty"`
+	// A JSON array of criteria to sort the result by.
+	// Each element of the array is a JSON object that specifies one ordering.
+	// The position in the array identifies the rank of an ordering, i.e., whether it is primary, secondary, etc.
+	Sorting []ReqSort `json:"sorting"`
+}
+
+type ResHistoryTaskInstance struct {
+	// The id of the task.
+	Id *string `json:"id"`
+	// The key of the process definition the task belongs to.
+	ProcessDefinitionKey *string `json:"processDefinitionKey"`
+	// The id of the process definition this task belongs to.
+	ProcessDefinitionId *string `json:"processDefinitionId"`
```

Comment:
`string` and so on in the file?

My idea was this:
If the API documentation says that the parameter may not be present in JSON, then ",omitempty" is added to the anatization, and "*" is added to the beginning of the type.

This approach allows JSON serialization to both exclude optional fields and pass them with an empty value.

Because an empty value is also a value.

Reply:
sorry for the delay. A little busy plus a little lazy. Yes you're correct though the API document doesn't mark if the field is optional or now.